dependency/dependency_memory.go

```diff
@@ -0,0 +1,74 @@
+package dependency
+
+import (
+	"sync"
+
+	"github.com/vacp2p/mvds/state"
+)
+
+// Verify that MessageDependency interface is implemented.
+var _ MessageDependency = (*memoryDependency)(nil)
+
+type memoryDependency struct {
+	sync.Mutex
+
+	dependents map[state.MessageID][]state.MessageID
+	dependencies map[state.MessageID]int
+
+}
+
+func NewMemoryDependency() *memoryDependency {
+	return &memoryDependency{
+		dependents: make(map[state.MessageID][]state.MessageID),
+		dependencies: make(map[state.MessageID]int),
+	}
+}
+
+func (md *memoryDependency) Add(msg, dependency state.MessageID) error {
+	md.Lock()
+	defer md.Unlock()
+	// @todo check it wasn't already added
+	md.dependents[dependency] = append(md.dependents[dependency], msg)
+	md.dependencies[msg] += 1
+	return nil
+}
+
+func (md *memoryDependency) Dependants(id state.MessageID) ([]state.MessageID, error) {
+	md.Lock()
+	defer md.Unlock()
+
+	return md.dependents[id], nil
+}
+
+func (md *memoryDependency) MarkResolved(msg state.MessageID, dependency state.MessageID) error {
+	md.Lock()
+	defer md.Unlock()
+
+	id := -1
+	for i, val := range md.dependents[dependency] {
```

Comment:
You can do this because you stop iterating immediately when an item is found.

```suggestion
	for _, item := range md.dependents[dependency] {
		if item != msg {
		    continue
		}
         	md.dependents[dependency] = remove(md.dependents[dependency], id)
		md.dependencies[msg] -= 1
		return nil
	}
	return nil
```